packages/helia/src/block-providers/trustless-gateway-block-provider.ts

```diff
@@ -0,0 +1,83 @@
+import { logger } from '@libp2p/logger'
+import type { BlockProvider } from '@helia/interface/blocks'
+import type { AbortOptions } from 'interface-store'
+import type { CID } from 'multiformats/cid'
+import type { ProgressEvent, ProgressOptions } from 'progress-events'
+
+const log = logger('helia:trustless-gateway-block-provider')
+
+export type TrustlessGatewayGetBlockProgressEvents =
+  ProgressEvent<'trustless-gateway:get-block:fetch', URL>
+
+/**
+ * A BlockProvider that accepts a list of trustless gateways that are queried
+ * for blocks. Individual gateways are randomly chosen.
+ */
+export class TrustedGatewayBlockProvider implements BlockProvider<
+ProgressOptions,
+ProgressOptions<TrustlessGatewayGetBlockProgressEvents>
+> {
+  private readonly gateways: URL[]
+
+  constructor (urls: string[]) {
+    this.gateways = urls.map(url => new URL(url.toString()))
+  }
```

Comment:
prevent users from having to url.toString() just for us to convert it to a url again later. The code works either way and gives consumers flexibility

```suggestion
  constructor (urls: (string | URL)[]) {
    this.gateways = urls.map(url => new URL(url.toString()))
  }
```

Reply:
I think the linter prefers `Array<string | URL>` to this form.

Reply:
I've updated the signature.